### ISSUE TITLE
`npm test` often fails on OS X

### ISSUE THREAD

**Issue by yuya-oc** [yuya-oc/electron-mattermost#5]
Windows and Ubuntu(CircleCI) are OK. Console shows an error: `RuntimeError: Couldn't connect to selenium server`.
However, the test succeeds sometimes.
```
  electron-mattermost
    ✓ should show settings.html when there is no config file (1298ms)
    ✓ should show index.html when there is config file (965ms)
    ✓ should upgrade v0 config file (1004ms)
    index.html
      ✓ should set src of webview from config file (1260ms)
      ✓ should set name of tab from config file (1174ms)
RuntimeError: Couldn't connect to selenium server 
    at elementIdDisplayed("0.5625486401841044-2") - isVisible.js:56:5 
      1) should show only the selected team
    settings.html
      ✓ should show index.thml when Cancel button is clicked (2591ms)
      ✓ should show index.thml when Save button is clicked (2592ms)

  settings.js
    ✓ should upgrade v0 config file


  8 passing (22s)
  1 failing

  1) electron-mattermost index.html should show only the selected team:
     Error: timeout of 10000ms exceeded. Ensure the done() callback is being called in this test.
```

**Comment by kaleposhobios** [comment]
I tried a lot of things with this but wasn't able to completely determine a root cause. Looking at the source error: https://github.com/webdriverio/webdriverio/blob/24e30dc6cb0c2429ac4e8863ef73eb874ef865dd/lib/utils/RequestHandler.js
It just seems like the selenium server (chromedriver) isn't running (to accept commands) at the time the test tries to send its test command. I'm wondering if maybe there is some occasional delay between when Chrome starts and when the chromedriver will actually accept commands for it.

In other words, I'm not exactly sure what the "proper" fix is. But I wonder if, as a workaround, it would work to put a delay after driver.init() before any other commands are called, which would give the Chrome app time to really launch and the chromedriver attach to it.

What do you think?

--Matt

**Comment by yuya-oc** [comment]
Thank you for analysis.

Yes, I had thought the delay is necessary. "WaitFor" commands may cause same issue, so I tried follwoing code. However, it also fails in my side.

```js
// Environment: OS X 10.11, Node.js v5.1.0
beforeEach(function(done){
  client.init().then(function(){
    setTimeout(done, 1000); // I tried some delay duration.
  });
});

it('should...', function(done){
  client
    .some_commands()
    ...
});
```

**Comment by kaleposhobios** [comment]
Good to know, thanks. I played around a bit more. One odd thing is, it should be equivalent to call:
    client.init().some_commands()
or
```
client.init().then(function () {
    client.some_commands()
});
```
These should be equivalent, right? I changed it just to try it, and I couldn't actually reproduce the failure after doing so. I'm assuming this is a coincidence. But, what do you think?

**Comment by yuya-oc** [comment]
Hmm...yes, I think they are equivalent. But it's odd as you said.
If your changes still don't reproduce the failure, would you show code whole?

**Comment by kaleposhobios** [comment]
I think it was probably a false lead but, just in case it really does make a difference, if you want to try it, each of my tests are now:
{{{
  it('should show settings.html when there is no config file', function(done) {
    client
      .init().then(function () {
        client.getUrl().then(function(url) {
          var p = path.parse(url);
          p.base.should.equal('settings.html');
        })
        .end().then(function() {
          done();
        });
      })

  });
}}}

**Comment by yuya-oc** [comment]
>> uncaught exception write EPIPE at objects.exports._errnoexception utils.js:865:11...

> Have you seen anything like this? The error stays on my screen but after 10 seconds (the timeout) another Electron application opens, so there are two open.

No, I haven't seen this. I had added client.end() in afterEach() to avoid that there were two open.

I pushed a new branch, **test-on-osx**. The failure still happens, but the result is interesting. I added console.log into checkVisility(). So, I can know the test had succeeded until first isVisible() command. 

```
  electron-mattermost
    ✓ should show settings.html when there is no config file (1295ms)
    ✓ should show index.html when there is config file (962ms)
    ✓ should upgrade v0 config file (942ms)
    index.html
      ✓ should set src of webview from config file (1215ms)
      ✓ should set name of tab from config file (1133ms)
[ true, false ]
RuntimeError: Couldn't connect to selenium server 
    at elementIdDisplayed("0.2587126281578094-2") - isVisible.js:56:5 
      1) should show only the selected team
    settings.html
      ✓ should show index.thml when Cancel button is clicked (2580ms)
      ✓ should show index.thml when Save button is clicked (2597ms)

  settings.js
    ✓ should upgrade v0 config file


  8 passing (30s)
  1 failing

  1) electron-mattermost index.html should show only the selected team:
     Error: timeout of 10000ms exceeded. Ensure the done() callback is being called in this test.
  



npm ERR! Test failed.  See above for more details.
```

As result, I think there is possible that the reason is the connection dropping. However, if it was true, I have no idea to suppress.

**Comment by kaleposhobios** [comment]
Hi again,

I have no idea if this will be helpful, but I did more investigation. I turned on chromedriver logs and based on those, it seems like the forEach loop to evaluate multiple matches, doesn't seem to be working. I'm not sure if this is an issue inside webdriverio or the test suite, though.

The reason I'm wondering this is: I turned on chromedriver debug logging by adding:
```
browser_test.js line 39: 
chromedriver = require('child_process').spawn('node_modules/chromedriver/lib/chromedriver/chromedriver', ['--url-base=wd/hub', '--port=9515', '--verbose', '--log-path=[absolute path to log file]']);
```
Then, I commented out most of the tests, leaving just the failing tests within describe('index.html'

When I then ran the tests, it failed on the very first one:
```
electron-mattermost
    index.html
RuntimeError: Couldn't connect to selenium server 
    at elementIdAttribute("0.4014562803786248-2", "src") - getAttribute.js:51:5 
      1) should set src of webview from config file
```
Based on the test, I would expect that Chromedriver would get two requests, one for the first tab and one for the second. But I looked in the Chromedriver log and don't see what I'd expect. I have pasted below the relevant output. Lines prefaced by [Matthew] are added by me, the rest are the log output:
```
[2.019][DEBUG]: DEVTOOLS RESPONSE Runtime.evaluate (id=11) {
   "result": {
      "type": "object",
      "value": {
         "status": 0,
         "value": [ {
            "ELEMENT": "0.4014562803786248-1"
         }, {
            "ELEMENT": "0.4014562803786248-2"
         } ]
      }
   },
   "wasThrown": false
}
[Matthew] Looks like it found the two elements of class mattermostView
[2.019][INFO]: Waiting for pending navigations...
[2.019][INFO]: Done waiting for pending navigations
[2.019][INFO]: RESPONSE FindElements [ {
   "ELEMENT": "0.4014562803786248-1"
}, {
   "ELEMENT": "0.4014562803786248-2"
} ]
[Matthew] Looks like it is now responding to the test suite, with those 2 elements
[2.038][INFO]: COMMAND GetElementAttribute {
   "id": "0.4014562803786248-2",
   "name": "src"
}
[Matthew] Ok, here is the command from the test suite for get src attribute of the 2nd element
[2.038][INFO]: Waiting for pending navigations...
[2.038][INFO]: Done waiting for pending navigations
[2.039][DEBUG]: DEVTOOLS COMMAND Runtime.evaluate (id=12) {
   "expression": "(function() { // Copyright (c) 2012 The Chromium Authors. All rights reserved.\n// Use of this source code is governed by a BSD-style license that can be\n// found in the LICENSE file.\n\n/**\n * Enum f...",
   "returnByValue": true
}
[2.063][DEBUG]: DEVTOOLS RESPONSE Runtime.evaluate (id=12) {
   "result": {
      "type": "object",
      "value": {
         "status": 0,
         "value": "http://example.com/team"
      }
   },
   "wasThrown": false
}
[2.063][INFO]: Waiting for pending navigations...
[2.063][INFO]: Done waiting for pending navigations
[2.063][INFO]: RESPONSE GetElementAttribute "http://example.com/team"
[Matthew] Ok, now it responded to the test suite with the value of the 2nd element
[11.205][INFO]: COMMAND Quit {

}
[Matthew] Now it got the quit command. But, where was the request for the src of the 1st element?
[11.256][INFO]: RESPONSE Quit
[11.256][DEBUG]: Log type 'driver' lost 71 entries on destruction
[11.256][DEBUG]: Log type 'browser' lost 0 entries on destruction
[11.346][INFO]: COMMAND InitSession {
[Matthew] Now beginning next test...
```
I saw the same thing multiple times. I also saw the case where there is no error thrown (I removed what seem to me to be irrelevant lines this time, to make it easier to see:
```
[4.423][INFO]: RESPONSE FindElements [ {
   "ELEMENT": "0.35136634879745543-1"
}, {
   "ELEMENT": "0.35136634879745543-2"
} ]
[4.436][INFO]: COMMAND IsElementDisplayed {
   "id": "0.35136634879745543-1"
}
[4.446][INFO]: RESPONSE IsElementDisplayed false
[4.446][INFO]: COMMAND IsElementDisplayed {
   "id": "0.35136634879745543-2"
}
 ```
So basically it seems like when it works, we can see the Chromedriver log clearly showing that it gets a request for the 1st element, then a request for the 2nd, and the test passes.

I don't know for sure that this is a bug in webdriverio, but I suspect it is. I know the error is "can't connect to driver" but, the Chromedriver log clearly shows a connection and request for the 2nd element. How come the request for the 1st item isn't showing up? Unless it's a temporary connection issue, maybe it's  bug with the foreach somewhere?

I'm sorry if this isn't helpful and wastes your time. But this definitely, to me, seems to point to a bug that's not in your test suite code?

**Comment by yuya-oc** [comment]
Thanks a lot. Don't worry, this was good time for the issue.
Fortunately(Unfortunately), there is no error on Linux. So the issue is not critical.

When I have time to study, I'll also try checking webdriverio.

**Comment by yuya-oc** [comment]
For now, I had modified tests to avoid multiple elements. Please reopen if it's necessary.
